src/Bicep.Core/Analyzers/Linter/Rules/UseRecentApiVersionRule.cs

```diff
@@ -0,0 +1,326 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using Bicep.Core.ApiVersions;
+using Bicep.Core.CodeAction;
+using Bicep.Core.Configuration;
+using Bicep.Core.Diagnostics;
+using Bicep.Core.Parsing;
+using Bicep.Core.Resources;
+using Bicep.Core.Semantics;
+using Bicep.Core.Syntax;
+using Bicep.Core.TypeSystem;
+
+namespace Bicep.Core.Analyzers.Linter.Rules
+{
+    public sealed class UseRecentApiVersionRule : LinterRuleBase
+    {
+        public new const string Code = "use-recent-api-version";
+        public const int MaxAllowedAgeInDays = 365 * 2;
+
+        // Debug/test switch: Pretend today is a different date
+        private DateTime today = DateTime.Today;
+
+        // Debug/test switch: Warn if the resource type or API version are not found (normally we don't
+        // give errors for these because Bicep always provides a warning about types not being available)
+        private bool warnNotFound = false;
+
+        public UseRecentApiVersionRule() : base(
+            code: Code,
+            description: CoreResources.UseRecentApiVersionRule_Description,
+            docUri: new Uri($"https://aka.ms/bicep/linter/{Code}"),
+            diagnosticStyling: DiagnosticStyling.Default)
+        {
+        }
+
+        public override void Configure(AnalyzersConfiguration config)
+        {
+            base.Configure(config);
+
+            // Today's date can be changed to enable testing/debug scenarios
+            string? testToday = this.GetConfigurationValue<string?>("test-today", null);
+            if (testToday is not null)
+            {
+                this.today = ApiVersionHelper.ParseDateFromApiVersion(testToday);
+            }
+
+            // Testing/debug: Warn if the resource type and/or API version are not found
+            bool debugWarnNotFound = this.GetConfigurationValue<bool>("test-warn-not-found", false);
+            this.warnNotFound = debugWarnNotFound == true;
+
+        }
+
+        public override string FormatMessage(params object[] values)
+        {
+            var resourceType = (string)values[0];
+            var reason = (string)values[1];
+            var acceptableVersions = (ApiVersion[])values[2];
+
+            var acceptableVersionsString = string.Join(", ", acceptableVersions.Select(v => v.Formatted));
+            return string.Format(CoreResources.UseRecentApiVersionRule_ErrorMessageFormat, resourceType)
+                + (" " + reason)
+                + (acceptableVersionsString.Any() ? " " + string.Format(CoreResources.UseRecentApiVersionRule_AcceptableVersions, acceptableVersionsString) : "");
+        }
+
+        override public IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel model)
+        {
+            var visitor = new Visitor(model, today, UseRecentApiVersionRule.MaxAllowedAgeInDays, warnNotFound);
```

Comment:
We'll need to exclude non-Az resources here as they won't be following the same apiverion structure, or may not have an apiversion specified at all. The easiest way to do this is to check the `IsAzResource` property on resource metadata.
https://github.com/Azure/bicep/blob/b32aa90d91f8bb928836c152f5ba09cd4cc891f6/src/Bicep.Core/Analyzers/Linter/Rules/UseProtectedSettingsForCommandToExecuteSecretsRule.cs#L42

It would probably also be worth adding a quick test to verify that this linter doesn't run on a non-az resource. Here's a sample you could use:
https://github.com/Azure/bicep/blob/b32aa90d91f8bb928836c152f5ba09cd4cc891f6/src/Bicep.Core.IntegrationTests/ExtensibilityTests.cs#L190-L205

Note that you'll need to explicitly enable the extensibility feature (`importsEnabled`) when you construct the features provider, as it is off by default:
https://github.com/Azure/bicep/blob/b32aa90d91f8bb928836c152f5ba09cd4cc891f6/src/Bicep.Core.IntegrationTests/ExtensibilityTests.cs#L23

Reply:
Done, thanks!

I wonder if this should be built into the API better, though, to encourage the right code.  Maybe TryGetResourceTypeReference should require passing in Az vs k8 vs whatever?